recirq/quantum_chess/quantum_board_test.py

```diff
@@ -1412,3 +1412,71 @@ def test_split_capture_with_failed_measurement_outcome(board):
         probs = b.get_probability_distribution(100)
         assert_prob_about(probs, u.square_to_bit("a2"), 1, atol=0.1)
         assert_prob_about(probs, u.square_to_bit("a3"), 0, atol=0.1)
+
+
+@pytest.mark.parametrize("board", ALL_CIRQ_BOARDS)
+def test_merge_to_fully_classical_position(board):
+    """Splits piece on d4 to b3 and c2, then merge back to fully classical position on a1."""
+    b = simulator(u.squares_to_bitboard(["d4"]))
+    b.reset_starting_states = True
+    assert b.perform_moves(
+        "d4^b3c2:SPLIT_JUMP:BASIC",
+        "b3c2^a1:MERGE_JUMP:BASIC",
+    )
+
+    assert b.is_classical()
+
+
+@pytest.mark.parametrize("board", ALL_CIRQ_BOARDS)
+def test_undo_to_start_after_measurement(board):
+    """Splits piece on f8 to d6 and h6. Piece on f4 then captures piece on d6. Then piece on h6 captures d6
+    Then does three undo moves to return to initial position."""
+    b = simulator(u.squares_to_bitboard(["f4", "f8", "f2"]))
+    b.reset_starting_states = True
+    initial_board = b.get_full_squares_bitboard()
+    assert b.perform_moves(
+        "f8^d6h6:SPLIT_JUMP:BASIC",
+        "f4^d6:JUMP:CAPTURE",
+        "f2^h6:JUMP:CAPTURE",
+    )
+
+    assert b.is_classical()
+    assert b.circuit == cirq.Circuit()
+
+    assert b.undo_last_move()
+    assert b.undo_last_move()
+    assert b.undo_last_move()
+
+    assert b.get_full_squares_bitboard() == initial_board
+
+
+@pytest.mark.parametrize("board", ALL_CIRQ_BOARDS)
+def test_quantum_capture_with_forced_measurement(board):
+    """Splits piece on b1 to a1 and c1. Piece on a1 then captures piece on a5."""
+    b = simulator(u.squares_to_bitboard(["b1", "a5"]))
+    b.reset_starting_states = True
+
+    b.perform_moves(
+        "b1^a1c1:SPLIT_JUMP:BASIC",
+        "a1^a5:JUMP:CAPTURE",
+    )
+
+    assert b.is_classical()
+    assert b.move_history != []
+    assert b.circuit == cirq.Circuit()
+    assert b.ancilla_count == 0
+
+
+@pytest.mark.parametrize("board", ALL_CIRQ_BOARDS)
+def test_measurement_without_fully_classical_position(board):
+    """Splits piece on d3 to d7 and h3, then splits piece on b5 to b7 and b3.
+    Piece on f5 then moves to c8. Piece on c8 then moves to h3, with failed measurement."""
```

Comment:
> Piece on c8 then moves to h3, with failed measurement

All moves here have the BASIC move variant, which means no measurement is performed (unless the move type is castling). CAPTURE  or EXCLUDED variants do measurements